brian2/codegen/runtime/numpy_rt/templates/spikegenerator.py_

```diff
@@ -1,12 +1,41 @@
-{# USES_VARIABLES { _spikespace, t, dt, neuron_index, spike_time } #}
+{# USES_VARIABLES { _spikespace, t, dt, neuron_index, spike_time, period, _lastindex } #}
 import numpy as np
 
 # TODO: We don't deal with more than one spike per neuron yet
-# Spikes since the preceding time step
-start = np.searchsorted({{spike_time}}, t-dt, side='right')
-stop = np.searchsorted({{spike_time}}, t, side='right')
-_indices = {{neuron_index}}[start:stop]
 
-_n_spikes = len(_indices)
+padding_before = t % period
+padding_after  = (t + dt) % period
+_n_spikes      = 0
+epsilon        = 1e-3*dt
+
+# We need some precomputed values that will be used during looping
+not_first_spike = {{_lastindex}}[0] > 0
+not_end_period  = abs(padding_after) > (dt - epsilon)
+
+# If there is a periodicity in the SpikeGenerator, we need to reset the lastindex 
+# when all spikes have been played and at the end of the period
+if not_first_spike and ({{spike_time}}[{{_lastindex}}[0] - 1] > padding_before):
+    {{_lastindex}}[0] = 0
+
+for _spike_time in {{spike_time}}[{{_lastindex}}[0]:]:
+    test = (_spike_time > padding_before) or (abs(_spike_time - padding_before) < epsilon)
+    if test:
+        break
+    {{_lastindex}}[0] += 1
```

Comment:
As far as I remember, those tests were because I had problem when I had
spikes, with a non zero period, that were falling in the first or the last
time-step. I wanted to be sure that all were played and played only once.
But if you say that they can be removed...

2015-01-21 16:32 GMT+01:00 Marcel Stimberg <notifications@github.com>:

> In brian2/codegen/runtime/numpy_rt/templates/spikegenerator.py_
> <https://github.com/brian-team/brian2/pull/339#discussion_r23306810>:
>
> > +epsilon        = 1e-3*dt
> > +
> > +# We need some precomputed values that will be used during looping
> > +not_first_spike = {{_lastindex}}[0] > 0
> > +not_end_period  = abs(padding_after) > (dt - epsilon)
> > +
> > +# If there is a periodicity in the SpikeGenerator, we need to reset the lastindex
> > +# when all spikes have been played and at the end of the period
> > +if not_first_spike and ({{spike_time}}[{{_lastindex}}[0] - 1] > padding_before):
> > +    {{_lastindex}}[0] = 0
> > +
> > +for _spike_time in {{spike_time}}[{{_lastindex}}[0]:]:
> > +    test = (_spike_time > padding_before) or (abs(_spike_time - padding_before) < epsilon)
> > +    if test:
> > +        break
> > +    {{_lastindex}}[0] += 1
>
> What are lines 20-24 for, actually (same for the C code)? Forwarding the
> _lastindex if the spike times after it are in fact still before the
> current time -- can this situation ever occur? It does not seem to be
> triggered in any of the tests and they all pass if the code is deleted...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/brian-team/brian2/pull/339/files#r23306810>.
>